### PR TITLE
feat(langchain): add support for langchain==0.1

### LINKED ISSUE #8212: Latest Langchain throws deprecated warnings
### Summary of problem
Upgrading to the latest langchain, we are now getting deprecated warnings

LangChainDeprecationWarning: Importing chat models from langchain is deprecated. Importing from langchain will no longer be supported as of langchain==0.2.0. Please import from langchain-community instead:\n\n`from langchain_community.chat_models import ChatOpenAI`

Vectorstores also throws this deprecated warning.

It looks like it is trying import the old code first, then if it can't find it, it imports the correct code. 

### Which version of dd-trace-py are you using?
2.5.1

### Which libraries and their versions are you using?
langchain - 0.1.4
langchain-community - 0.0.14
langchain-core - 0.1.16
langchain-openai - 0.0.4

<details>
  <summary>Upgrading to the latest langchain, we are now getting deprecated warnings</summary>
  
</details>

### How can we reproduce your problem?
Use the latest langchain with ddtrace you will get deprecated warnings. 

### What is the result that you expected?
Update to the recommended calls so the warnings go away.

### PR DESCRIPTION
**Pull request by sabrenner**
Fixes #8212.

### For Reviewers

**NOTE: Please disregard the LOC count in this PR, 90% of the lines of code changed are cassette test data / moved files / snapshots files / requirements lockfiles.**
The important files to review are:
- `ddtrace/contrib/langchain/patch.py`: version gating which methods to patch
- `releasenotes/notes/feat-support-langchain-0-1-0d8e0ddd6248c4ed.yaml` for release note
- `tests/contrib/langchain/test_langchain_community.py` for testing langchain >= 0.1
- `tests/contrib/langchain/test_langchain_patch.py` for testing which methods are traced

## Change Summary

This PR adds support for `langchain>=0.1`, which has deprecated multiple traced methods from `langchain<=0.0.354` for removal in the upcoming `langchain==0.2.0` release. Whereas the older version of langchain was all contained in one library `langchain`, the new version has split its library into separate subpackages:
- `langchain_core`: Core base classes for llms, chat models
- `langchain_community`: Community subpackage that contains integrations for different libraries (i.e. `cohere, anthropic, etc...`)
- `langchain_<partner_name>`: Individual subpackages for each partner integration such as `langchain_openai`, `langchain_pinecone`. It appears that the ultimate goal is for all of the code in `langchain_community` to get extracted to these partner subpackages.
- `langchain`: The old library which contains still a lot of base methods such as Chains.

tldr; the functionality in this PR adds version gating the langchain integration to patch the correct methods/import paths with the correct traced functions. No functionality has been added.

### Risks
Currently a design flaw of the langchain integration is that it keeps a static list of embeddings/vectorstore class names to then patch `langchain_community.embeddings/vectorstores.<CLASS_NAME>`. However with langchain's apparent push for moving integrations to their individual partner subpackages, this will break our patching for embeddings/vectorstores since now we have to patch each subpackage individually. Currently only pinecone/openai (which our integration specifically patches) and elasticsearch/vertexAI (not yet patched) have their own subpackages, so the risk is low since there will be no breaking changes (and [Langchain promises no more breaking changes on minor releases](https://blog.langchain.dev/week-of-1-22-24-langchain-release-notes/#:~:text=No%20more%20breaking%20changes%20on%20a%20minor%20version%20release.)).

Our proposed approach of resolving this issue is to offer limited tracing support for embeddings/vectorstores with individual subpackages (openai/pinecone for now) at first, then gradually expand our support once the others become available. We will likely need to redesign/refactor the integration in that case.

## Testing
This PR adds a separate venv for `langchain-community/langchain-core/langchain-openai/langchain-pinecone`, and only tests with python 3.10 and over. The reason is that python 3.9 and under requires a separate set of test cassette files, and that is an unnecessary number of files/tests. 

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/contrib/langchain/patch.py [hunk 1]
```diff
@@ -51,6 +64,15 @@ def get_version():
     return getattr(langchain, "__version__", "")
 
 
+# After 0.1.0, implementation split into langchain, langchain_community, and langchain_core.
+# We need to check the version to determine which module to wrap, to avoid deprecation warnings
+# ref: https://github.com/DataDog/dd-trace-py/issues/8212
+LANGCHAIN_VERSION = parse_version(get_version())
+SHOULD_USE_LANGCHAIN_COMMUNITY = LANGCHAIN_VERSION >= (0, 1, 0)
+BASE_LANGCHAIN_MODULE = langchain_community if SHOULD_USE_LANGCHAIN_COMMUNITY else langchain
```
Comment (mabdinur) [inline ddtrace/contrib/langchain/patch.py:72]: BASE_LANGCHAIN_MODULE allows us to patch some `langchain_community` and `langchain` objects in the same way. Over the next few versions the structure of both packages might start to differ significantly.  Would it be easier to maintain an integration with minimal coupling? 

I am fine with doing this as a follow up

File: ddtrace/contrib/langchain/patch.py [hunk 2]
```diff
@@ -51,6 +64,15 @@ def get_version():
...
+BASE_LANGCHAIN_MODULE_NAME = getattr(BASE_LANGCHAIN_MODULE, "__name__", "langchain")
```
Comment (mabdinur) [inline ddtrace/contrib/langchain/patch.py:73]: nit: `__name__` should always be defined but I guess the additional safely doesn't hurt.
```suggestion
BASE_LANGCHAIN_MODULE_NAME = BASE_LANGCHAIN_MODULE.__name__
```

File: ddtrace/contrib/langchain/patch.py [hunk 3]
```diff
@@ -154,7 +176,11 @@ def traced_llm_generate(langchain, pin, func, instance, args, kwargs):
                 span.set_tag_str("langchain.request.%s.parameters.%s" % (llm_provider, param), str(val))
 
         completions = func(*args, **kwargs)
-        if isinstance(instance, langchain.llms.OpenAI):
+        if (
+            isinstance(instance, BASE_LANGCHAIN_MODULE.llms.OpenAI)
+            or langchain_openai
+            and isinstance(instance, langchain_openai.OpenAI)
```
Comment (mabdinur) [inline ddtrace/contrib/langchain/patch.py:182]: Question `if True or False and True` will evaluate to `True` and . Is this the expected behavior? Do we need brackets here?
```suggestion
            isinstance(instance, BASE_LANGCHAIN_MODULE.llms.OpenAI)
            or (langchain_openai
            and isinstance(instance, langchain_openai.OpenAI))
```

File: ddtrace/contrib/langchain/patch.py [hunk 4]
```diff
@@ -677,41 +717,83 @@ def patch():
...
-    wrap("langchain", "chat_models.base.BaseChatModel.generate", traced_chat_model_generate(langchain))
-    wrap("langchain", "chat_models.base.BaseChatModel.agenerate", traced_chat_model_agenerate(langchain))
-    wrap("langchain", "chains.base.Chain.__call__", traced_chain_call(langchain))
-    wrap("langchain", "chains.base.Chain.acall", traced_chain_acall(langchain))
+    if SHOULD_USE_LANGCHAIN_COMMUNITY:
+        from langchain.chains.base import Chain  # noqa:F401
+        from langchain_community import embeddings  # noqa:F401
+        from langchain_community import vectorstores  # noqa:F401
+        from langchain_core.language_models.chat_models import BaseChatModel  # noqa:F401
+        from langchain_core.language_models.llms import BaseLLM  # noqa:F401
```
Comment (mabdinur) [inline ddtrace/contrib/langchain/patch.py:725]: Do we need these imports? If they are not used in patch should we just move them to `unpatch`?

File: releasenotes/notes/feat-support-langchain-0-1-0d8e0ddd6248c4ed.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    langchain:  This introduces support for ``langchain==0.1.0``. Note that this drops tracing support for deprecated
+    operations in ``langchain==0.1.0`` in favor of recommended alternative methods by [langchain](https://python.langchain.com/docs/changelog/core).
```
Comment (mabdinur) [inline releasenotes/notes/feat-support-langchain-0-1-0d8e0ddd6248c4ed.yaml:5]: Markdown and rst links have different formats. Also I cleaned up the release note a little bit
```suggestion
    langchain operations. Please follow langchain upgrade `guide <https://python.langchain.com/docs/changelog/core>`_ to enable full tracing support.
```

File: releasenotes/notes/feat-support-langchain-0-1-0d8e0ddd6248c4ed.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    This includes:
+    - ``langchain.chain.Chain.run/arun`` with ``langchain.chain.Chain.invoke/ainvoke``
+    - ``langchain.embeddings.openai.OpenAIEmbeddings.embed_documents`` with ``langchain_openai.OpenAIEmbeddings.embed_documents``
+    - ``langchain.vectorstores.pinecone.Pinecone.similarity_search`` with ``langchain_pinecone.PineconeVectorStore.similarity_search``
```
Comment (mabdinur) [inline releasenotes/notes/feat-support-langchain-0-1-0d8e0ddd6248c4ed.yaml:9]: Can we move this to `ddtrace.contrib.langchain.__init__.py? We can then link the langchain docs here (l[ink to rst anchor docs](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#use-a-custom-anchor)). We do this in a few release notes so there should be examples

File: tests/contrib/langchain/conftest.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+def langchain_community(ddtrace_global_config, ddtrace_config_langchain, mock_logs, mock_metrics):
+    global_config = default_global_config()
+    global_config.update(ddtrace_global_config)
+    with override_global_config(global_config):
+        with override_config("langchain", ddtrace_config_langchain):
+            # ensure that mock OpenAI API key is passed in
+            os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY", "<not-a-real-key>")
+            os.environ["COHERE_API_KEY"] = os.getenv("COHERE_API_KEY", "<not-a-real-key>")
+            os.environ["HUGGINGFACEHUB_API_TOKEN"] = os.getenv("HUGGINGFACEHUB_API_TOKEN", "<not-a-real-key>")
+            os.environ["AI21_API_KEY"] = os.getenv("AI21_API_KEY", "<not-a-real-key>")
```
Comment (mabdinur) [inline tests/contrib/langchain/conftest.py:105]: We should use `override_env` here. Setting the envars in this fixture could leak to other tests

File: tests/contrib/langchain/conftest.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+
+@pytest.fixture
+def langchain_openai(ddtrace_global_config, ddtrace_config_langchain, mock_logs, mock_metrics):
+    global_config = default_global_config()
+    global_config.update(ddtrace_global_config)
+    with override_global_config(global_config):
+        with override_config("langchain", ddtrace_config_langchain):
+            # ensure that mock OpenAI API key is passed in
+            os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY", "<not-a-real-key>")
```
Comment (mabdinur) [inline tests/contrib/langchain/conftest.py:123]: overrride env

File: tests/contrib/langchain/test_langchain.py [hunk 9]
```diff
@@ -407,7 +282,7 @@ def test_openai_chat_model_sync_call_39(langchain, request_vcr):
 @pytest.mark.skipif(sys.version_info < (3, 10, 0), reason="Python 3.10+ specific test")
 @pytest.mark.snapshot(
     token="tests.contrib.langchain.test_langchain.test_openai_chat_model_generate",
-    ignores=["metrics.langchain.tokens.total_cost"],
+    ignores=["metrics.langchain.tokens.total_cost", "resource"],
```
Comment (mabdinur) [inline tests/contrib/langchain/test_langchain.py:285]: We do we need to ignore resource? Should we create multiple variants for this snapshot tests? Is this reasonable?

File: tests/contrib/langchain/test_langchain.py [hunk 10]
```diff
@@ -501,7 +374,6 @@ async def test_openai_chat_model_async_stream(langchain, request_vcr):
         await chat.agenerate([[langchain.schema.HumanMessage(content="What is the secret Krabby Patty recipe?")]])
 
 
-@flaky(1735812000)
```
Comment (mabdinur) [inline tests/contrib/langchain/test_langchain.py:504]: LET's GOOOO. No more flaky

File: tests/contrib/langchain/test_langchain.py [hunk 11]
```diff
@@ -979,7 +775,7 @@ def test_chain_logs(langchain, ddtrace_config_langchain, request_vcr, mock_logs,
                     "service": "",
                     "status": "info",
                     "ddtags": "env:,version:,langchain.request.provider:openai,langchain.request.model:text-davinci-003,langchain.request.type:llm,langchain.request.api_key:...key>",  # noqa: E501
-                    "dd.trace_id": str(llm_span.trace_id),
+                    "dd.trace_id": hex(llm_span.trace_id)[2:],
```
Comment (mabdinur) [inline tests/contrib/langchain/test_langchain.py:778]: nit: This will ensure you will always get zero padded trace_ids (and without the 0x)
```suggestion
                    "dd.trace_id": f"{llm_span.trace_id:0{16}x}",
```

**Review by mabdinur — COMMENTED** [review]
I like the idea of introducing support for v0.1 with minimal changes. After this PR is merged it would be nice to see a complete refactor where langchain `v0.0` and `v0.1` integrations have distinct patching but use common helper methods. This will allow use to better support future upgrades. Right now it's a bit hard to debug and maintain. 

I left some nits but overall the PR looks good to me

File: ddtrace/contrib/langchain/patch.py (hunk 3 above)
Comment (Yun-Kim) [reply ddtrace/contrib/langchain/patch.py:182]: Great catch!

File: ddtrace/contrib/langchain/patch.py (hunk 4 above)
Comment (Yun-Kim) [reply ddtrace/contrib/langchain/patch.py:725]: The last two you're right I can just remove them, but the chain/embeddings/vectorstore classes need to be imported first. This is because langchain doesn't allow wrapping directly from root, so we have to import the base classes first before wrapping.

File: tests/contrib/langchain/test_langchain.py (hunk 9 above)
Comment (Yun-Kim) [reply tests/contrib/langchain/test_langchain.py:285]: The resource name is the class path, which is different for langchain 0.0 vs 0.1 😠

File: tests/contrib/langchain/test_langchain.py (hunk 11 above)
Comment (Yun-Kim) [reply tests/contrib/langchain/test_langchain.py:778]: So this is the same as `hex(trace_id)[2:]`? Will the zero padding ever produce a different result?

File: ddtrace/contrib/langchain/patch.py (hunk 1 above)
Comment (Yun-Kim) [reply ddtrace/contrib/langchain/patch.py:72]: Yes, this is true. Will add this as a next step for the planned refactor work. Thanks!

File: ddtrace/contrib/langchain/patch.py [hunk 12]
```diff
@@ -63,8 +81,7 @@ def get_version():
 )
 
 
-def _extract_model_name(instance):
-    # type: (langchain.llm.BaseLLM) -> Optional[str]
+def _extract_model_name(instance: Any) -> Optional[str]:
     """Extract model name or ID from llm instance."""
```
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/langchain/patch.py:85]: Because Langchain offers support for multiple LLM providers (ex: openai, anthropic, cohere, huggingface, ...) there are probably as many unique model names as those llm providers host, but should still be bounded

**Review by lievan — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
